bindings/imap-codec-python/imap_codec.pyi

```diff
@@ -1,4 +1,156 @@
+        """
+
+    @staticmethod
+    def encode(authenticate_data: dict) -> Encoded:
+        """
+        Encode authenticate data line into fragments.
+
+        :param bytes: Given authenticate data line
+        :return: `Encoded` type holding fragments of encoded authenticate data line
+        """
+
+class CommandCodec:
+    """
+    Codec for commands.
+    """
+
+    @staticmethod
+    def decode(bytes: bytes) -> Tuple[bytes, dict]:
+        """
+        Decode command from given bytes.
+
+        :param bytes: Given bytes
+        :return: Tuple of remaining bytes and decoded command
+        """
+
+    @staticmethod
+    def encode(command: dict) -> Encoded:
+        """
+        Encode command into fragments.
+
+        :param bytes: Given command
+        :return: `Encoded` type holding fragments of encoded command
+        """
+
+class ResponseCodec:
+    """
+    Codec for responses.
+    """
+
+    @staticmethod
+    def decode(bytes: bytes) -> Tuple[bytes, dict]:
+        """
+        Decode response from given bytes.
+
+        :param bytes: Given bytes
+        :return: Tuple of remaining bytes and decoded response
+        """
+
+    @staticmethod
+    def encode(response: dict) -> Encoded:
+        """
+        Encode response into fragments.
+
+        :param bytes: Given response
+        :return: `Encoded` type holding fragments of encoded response
+        """
+
+class IdleDoneCodec:
+    """
+    Codec for idle dones.
+    """
+
+    @staticmethod
+    def decode(bytes: bytes) -> Tuple[bytes, Tuple[()]]:
+        """
+        Decode idle done from given bytes.
+
+        :param bytes: Given bytes
+        :return: Tuple of remaining bytes and decoded idle done
+        """
+
+    @staticmethod
+    def encode(idle_done: Tuple[()]) -> Encoded:
+        """
+        Encode idle done into fragments.
+
+        :param bytes: Given idle done
+        :return: `Encoded` type holding fragments of encoded idle done
+        """
+
+class DecodeError(Exception):
+    """
+    Error during decoding.
+    """
+
+class DecodeFailed(DecodeError):
+    """
+    "Failed" error during decoding:
+    Decoding failed.
+    """
+
+class DecodeIncomplete(DecodeError):
+    """
+    "Incomplete" error during decoding:
+    More data is needed.
+    """
+
+class DecodeLiteralFound(DecodeError):
+    """
+    "LiteralFound" error during decoding:
+    The decoder stopped at the beginning of literal data.
+    """
+
+class Encoded:
+    """
+    An encoded message.
+
+    This struct facilitates the implementation of IMAP client- and server implementations by
+    yielding the encoding of a message through [`Fragment`]s. This is required, because the usage of
+    literals (and some other types) may change the IMAP message flow. Thus, in many cases, it is an
+    error to just "dump" a message and send it over the network.
+    """
+
+    def __iter__(self) -> "Encoded": ...
```

Comment:
I was assuming needing a string here, because `Encoded` is not yet defined as a type (forward-reference), which would fail at function definition time:
```
Traceback (most recent call last):
  File "...\bindings\imap-codec-python\imap_codec.pyi", line 26, in <module>
    class Encoded:
  File "...\bindings\imap-codec-python\imap_codec.pyi", line 36, in Encoded
    def __iter__(self) -> Encoded: ...
                          ^^^^^^^
NameError: name 'Encoded' is not defined
```

While this assumption was generally correct, I actually never tried running the *.pyi file through the Python interpreter. When doing so, I noticed that the order of classes was also incorrect (using `Encoded` as return value before its class definition), which I now cleaned up. As it turns out, interface stub files are usually only parsed by tools like mypy or language servers, but I like the idea that the file is actually parsable by the interpreter as well.

However, there is a different way of dealing with this as defined by [PEP 563 – Postponed Evaluation of Annotations](https://peps.python.org/pep-0563/). The idea is that annotations (e.g. type hints) are no longer evaluated during function definiton time by the interpreter which prevents the `NameError` shown above. This alternative parsing behavior is available since Python 3.7 through `from __future__ import annotations`. There do not seem to be any disadvantages in using this, so I added this as well and removed the quotes. (Actually, the reordering was unnecessary now as well, but I like the consistent order through all files.)

Thanks for sending me down this rabbit hole! 😄

Reply:
Woah, thanks for the explanation. As an ousider to Python I didn't expect that this is so complicated.